### PR TITLE
ARROW-9790: [Rust][Parquet] Fix PrimitiveArrayReader boundary conditions

### PR DESCRIPTION
**Pull request by alamb**
When I was reading a parquet file into `RecordBatches` using `ParquetFileArrowReader` that had row groups that were 100,000 rows in length with a batch size of 60,000, after reading 300,000 rows successfully, I started seeing this error

```
 ParquetError("Parquet error: Not all children array length are the same!")
```

Upon investigation, I found that when reading with `ParquetFileArrowReader`, if the parquet input file has multiple row groups, and if a batch happens to end at the end of a row group for Int or Float, no subsequent row groups are read

Visually:

```
+-----+
| RG1 |
|     |
+-----+  <-- If a batch ends exactly at the end of this row group (page), RG2 is never read
+-----+
| RG2 |
|     |
+-----+
```

I traced the issue down to a bug in `PrimitiveArrayReader` where it mistakenly interprets reading `0` rows from a page reader as being at the end of the column.

This bug appears *not* to be present in the initial implementation #5378 -- FYI @andygrove  and @liurenjie1024 (the test harness in this file is awesome, btw), but was introduced in https://github.com/apache/arrow/pull/7140. I will do some more investigating to ensure the test case described in that ticket is handled

### CODE REVIEW

File: rust/parquet/src/arrow/array_reader.rs [hunk 1]
```diff
@@ -136,10 +136,8 @@ impl<T: DataType> ArrayReader for PrimitiveArrayReader<T> {
         while records_read < batch_size {
             let records_to_read = batch_size - records_read;
 
+            // NB can be 0 if at end of page
             let records_read_once = self.record_reader.read_records(records_to_read)?;
-            if records_read_once == 0 {
```
Comment (alamb) [inline rust/parquet/src/arrow/array_reader.rs:140]: The case of `0` rows being read is handled in the `if records_read_once < records_to-read` clause below -- namely in this case the code needs to try and get the next page of data from the page reader.

File: rust/parquet/src/arrow/arrow_reader.rs [hunk 2]
```diff
@@ -304,6 +304,45 @@ mod tests {
         >(2, 100, 2, message_type, 15, 50, converter);
     }
 
+    #[test]
```
Comment (alamb) [inline rust/parquet/src/arrow/arrow_reader.rs:307]: Both these tests fail without the changes in this PR.

I don't like the copy/paste nature of these tests and I plan a minor PR building on this one proposing how to remove the duplication and make the tests easier to read.

**Review by andygrove — APPROVED** [review]
LGTM

**Comment by alamb** [comment]
As the code that I removed in this PR  added to support empty pages in 8a61570a8bd64ae6f6e47ce8efd2287c3c2feded (but it didn't seem to have any tests), I was worried that this deletion would cause some sort of reversion in behavior. To ensure it didn't, I added a test for empty pages in commit cfdea8349a8f41229a4c8fb2cfddec3bb11df112

To try and ensure that the test covers the empty pages case, I and ran the new test on commit ed1f771dccdde623ce85e212eccb2b573185c461 (the one right before 8a61570a8bd64ae6f6e47ce8efd2287c3c2feded) and the test fails, thus leading me to conclude that the code removed in this PR is unnecessary and zero page parquet files are still supported.

Log of what I did:

Get commit right before support for empty pages:
```
alamb@MacBook-Pro rust % git checkout ed1f771dccdde623ce85e212eccb2b573185c461
git checkout ed1f771dccdde623ce85e212eccb2b573185c461
M	testing
Note: switching to 'ed1f771dccdde623ce85e212eccb2b573185c461'.
...
HEAD is now at ed1f771dc ARROW-8717: [CI][Packaging] Add build dependency on boost to homebrew
```

Apply new test for no pages:

```
alamb@MacBook-Pro rust % git cherry-pick cfdea8349a8f41229a4c8fb2cfddec3bb11df112
git cherry-pick cfdea8349a8f41229a4c8fb2cfddec3bb11df112
Auto-merging rust/parquet/src/arrow/array_reader.rs
[detached HEAD ab3bcdd8f] Add a test for reading empty pages
 Date: Wed Aug 19 10:16:31 2020 -0400
 1 file changed, 57 insertions(+), 1 deletion(-)
```

Then I had to edit the test a little to use `.len` instead of `.is_empty` (which did not exist in the old commit):

```
diff --git a/rust/parquet/src/arrow/array_reader.rs b/rust/parquet/src/arrow/array_reader.rs
index d70619518..54c929c05 100644
--- a/rust/parquet/src/arrow/array_reader.rs
+++ b/rust/parquet/src/arrow/array_reader.rs
@@ -967,7 +967,7 @@ mod tests {

         // expect no values to be read
         let array = array_reader.next_batch(50).unwrap();
-        assert!(array.is_empty());
+        assert!(array.len() == 0);
     }

     #[test]
```

And then I ran the test, and it fails:

```
cd arrow/rust/parquet && PARQUET_TEST_DATA=`pwd`/../../cpp/submodules/parquet-testing/data ARROW_TEST_DATA=`pwd`/../../testing/data cargo test
...
---- arrow::array_reader::tests::test_primitive_array_reader_empty_pages stdout ----
thread 'arrow::array_reader::tests::test_primitive_array_reader_empty_pages' panicked at 'called `Result::unwrap()` on an `Err` value: General("Can\'t build array without pages!")', parquet/src/arrow/array_reader.rs:962:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by andygrove** [comment]
@sunchao Could you review as well ?

File: rust/parquet/src/arrow/array_reader.rs [hunk 3]
```diff
@@ -1000,6 +998,33 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_primitive_array_reader_empty_pages() {
```
Comment (alamb) [inline rust/parquet/src/arrow/array_reader.rs:1002]: This is a test for the functionality added in https://github.com/apache/arrow/commit/8a61570a8bd64ae6f6e47ce8efd2287c3c2feded / https://github.com/apache/arrow/pull/7140

**Comment by sunchao** [comment]
> @sunchao Could you review as well ?

Thanks for pinging! will do today.

File: rust/parquet/src/arrow/arrow_reader.rs (hunk 2 above)
Comment (alamb) [reply rust/parquet/src/arrow/arrow_reader.rs:307]: https://github.com/apache/arrow/pull/8009

File: rust/parquet/src/arrow/array_reader.rs (hunk 1 above)
Comment (sunchao) [reply rust/parquet/src/arrow/array_reader.rs:140]: +1

**Review by sunchao — APPROVED** [review]
LGTM. Thanks @alamb .

**Pull request closed without merging**